foundations/src/telemetry/tokio/mod.rs

```diff
@@ -0,0 +1,364 @@
+//!
+//! ## Example
+//! ```no_run
+//! # use std::thread;
+//! # use std::time::Duration;
+//! # use foundations::telemetry::tokio::RuntimeMonitor;
+//! // create and monitor 8 runtimes spawned on other threads, then poll in the background from this one
+//! let mut monitor = RuntimeMonitor::new();
+//!
+//! for i in 0..8 {
+//!     let r = tokio::runtime::Builder::new_current_thread().enable_all().build().unwrap();
+//!     
+//!     monitor.add_runtime(None, Some(i), r.handle());
+//!
+//!     thread::spawn(move || r.block_on(async {
+//!         loop {
+//!             println!("Poll");
+//!             tokio::time::sleep(Duration::from_secs(1)).await;
+//!         }
+//!     }));
+//! }
+//!
+//! loop {
+//!     monitor.record_sample();
+//!
+//!     // record metrics roughly twice a second
+//!     std::thread::sleep(Duration::from_secs_f32(0.5));
+//! }
+//! ```
+
+use crate::telemetry::metrics::{metrics, Counter, Gauge};
+use std::sync::atomic::Ordering;
+use std::sync::Arc;
+use tokio::runtime::{Handle, RuntimeMetrics};
+
+/// Monitors a set of runtimes, allowing metrics to be periodically collected for each tracked runtime.
+#[derive(Default)]
+pub struct RuntimeMonitor {
+    runtimes: Vec<RuntimeHandle>,
+}
+
+impl RuntimeMonitor {
+    /// Construct a new runtime monitor which initially monitors no runtimes.
+    pub fn new() -> Self {
+        Self {
+            runtimes: Vec::new(),
+        }
+    }
+
+    /// Add a runtime to the monitor, optionally with a name and/or id in case you are monitoring multiple runtimes.
+    ///
+    /// Runtimes should be uniquely identifiable by both label and id.
+    pub fn add_runtime(
+        &mut self,
+        runtime_name: Option<Arc<str>>,
+        runtime_id: Option<usize>,
+        handle: &Handle,
+    ) {
+        if self.runtimes.iter().any(|handle| {
+            (handle.runtime_name.is_some() != runtime_name.is_some())
+                || (handle.runtime_id.is_some() != runtime_id.is_some())
+        }) {
+            panic!("If you specify a runtime name or ID for one runtime, that setting must be specified for all");
+        }
+
+        if self.runtimes.iter().any(|handle| {
+            (handle.runtime_name.is_some() && handle.runtime_name == runtime_name)
+                || (handle.runtime_id.is_some() && handle.runtime_id == runtime_id)
+        }) {
+            panic!("Any runtime names or IDs provided to the RuntimeMonitor must be unique");
+        }
+
+        self.runtimes.push(RuntimeHandle {
+            runtime_name,
+            runtime_id,
+            handle: handle.clone(),
+        });
+    }
+
+    /// Record a sample of runtime metrics for each tracked runtime.
+    pub fn record_sample(&mut self) {
+        self.runtimes
+            .iter_mut()
+            .for_each(RuntimeHandle::record_sample)
+    }
+}
+
+struct RuntimeHandle {
```

Comment:
Let's split monitor, this and metrics definitions into separate modules

Reply:
No objections to this on my end.